torch/nn/quantized/modules/functional_modules.py

```diff
@@ -0,0 +1,76 @@
+import torch
+from torch._ops import ops
+
+
+_VALID_OP_NAMES = {
+    'add',
+    'cat'
+}
+
+class FloatFunctional(torch.nn.Module):
+    def __init__(self, op_name):
```

Comment:
why do you need to initialize with op_name? I think it's much nicer API if you don't have to specify it - then your __init__ would be very simple. You could add some handling inside to ensure that only one method is ever called, but I'd say even that is optional.

Reply:
I had suggested this to protect from users doing
```
 # in init
self. f1 = nn.quantize.functionalwrapper()

def forward(self ,x, y):
   self.f1.add(x,y)
   self.f1.cat([x,x])
```
This should be an error: We cannot have f1 be used for multiple operations, as the observers have to be per instance.

Also, explicitly specifying the operation at init is consistent with how modules are used. A module does a fixed operation, defined at init. With the wrapper, it is hard to determine what the operations are looking at the init.

Reply:
I am not sure what to do here -- we had a long discussion about that, so we need to make a decision soon if we need that or not.

Reply:
I'd vote for not specifying it in constructor and @soumith was voting for the same. It's strictly more lines of code and still not a complete protection as I can mix up two different adds together. The benefit of not specifying it in constructor is that one can build a nice wrapper like (for eager only, but still cute):

```
def __init__(...):
  self.stats = [nn.quantize.functionalwrapper() for i in range(10)]

def forward():
  counter = Counter()
  z = self.stats[counter.next()].add(x,y)
  return self.stats[counter.next()].cat([z,z])
```

We shouldn't provide it as default, but it feels more flexible.

We could remember the first op being invoked in a field on python side and re-check it next time. But it'd be interfere with torchscript a bit today.

Reply:
Agree with the comment about incomplete protection, but isn't some protection better than none?.  But do not understand the benefit of the wrapper. One could always do:
```
def __init__(...):
  opList = ['add', 'cat', ....]
  self.stats = [nn.quantize.functionalwrapper(opList[i]) for i in range(len(opList)]

def forward():
  counter = Counter()
  z = self.stats[counter.next()].add(x,y)
  return self.stats[counter.next()].cat([z,z])
```
@dzhulgakov : If you are convinced about having no argument in the constructor, lets go with it.

Reply:
I am ok with either implementation. However, to clarify the purpose of the name in the constructor is to make sure that the functionalwrapper is used as intended in the constructor. I can see benefits of it from usage safety. But at the same time, I think this is getting into domain of over-protecting the user.

What I propose is, why don't we submit one or another, and discuss it in more details?

Reply:
I'll use @soumith as a tie-breaker - he voted for no name when we spoke earlier.

@raghuramank100 - in your example, it's still needed to align function types in two places (for example if one comments out some code)

Reply:
Yes, we still need to align function names in two places.  Lets go with removing the name as an argument. A todo will be to have torchscript ensure that observer instantiations are unique.